src/Sylius/Bundle/CartBundle/Controller/CartItemController.php

```diff
@@ -53,49 +36,79 @@ public function addAction(Request $request)
 
-        $cart = $this->getCurrentCart();
-        $emptyItem = $this->newResourceFactory->create($configuration, $this->factory);
+        $this->isGrantedOr403($configuration, ResourceActions::CREATE);
+        $newResource = $this->newResourceFactory->create($configuration, $this->factory);
 
-        $eventDispatcher = $this->getEventDispatcher();
+        $itemQuantityModifier = $this->get('sylius.order_item_quantity_modifier');
+        $itemQuantityModifier->modify($newResource, 1);
 
-        try {
-            $item = $this->getResolver()->resolve($emptyItem, $request);
-        } catch (ItemResolvingException $exception) {
-            // Write flash message
-            $eventDispatcher->dispatch(SyliusCartEvents::ITEM_ADD_ERROR, new FlashEvent($exception->getMessage()));
+        $form = $this->resourceFormFactory->create($configuration, $newResource);
 
-            return $this->redirectAfterAdd($configuration);
-        }
+        if ($request->isMethod('POST') && $form->submit($request)->isValid()) {
+            $newResource = $form->getData();
 
-        $event = new CartItemEvent($cart, $item);
+            $event = $this->eventDispatcher->dispatchPreEvent(ResourceActions::CREATE, $configuration, $newResource);
 
-        // Update models
-        $eventDispatcher->dispatch(SyliusCartEvents::ITEM_ADD_INITIALIZE, $event);
-        $eventDispatcher->dispatch(SyliusCartEvents::CART_CHANGE, new GenericEvent($cart));
-        $eventDispatcher->dispatch(SyliusCartEvents::CART_SAVE_INITIALIZE, $event);
+            if ($event->isStopped() && !$configuration->isHtmlRequest()) {
+                throw new HttpException($event->getErrorCode(), $event->getMessage());
+            }
+            if ($event->isStopped()) {
+                $this->flashHelper->addFlashFromEvent($configuration, $event);
 
-        // Write flash message
-        $eventDispatcher->dispatch(SyliusCartEvents::ITEM_ADD_COMPLETED, new FlashEvent());
+                return $this->redirectHandler->redirectToIndex($configuration, $newResource);
+            }
 
-        return $this->redirectAfterAdd($configuration);
-    }
+            $cart = $this->getCurrentCart();
 
-    /**
-     * Redirect to specific URL or to cart.
-     *
-     * @param Request $request
-     *
-     * @return RedirectResponse
-     */
-    private function redirectAfterAdd(RequestConfiguration $configuration)
-    {
-        $request = $configuration->getRequest();
+            $isItemInCart = false;
+            foreach ($cart->getItems() as $item) {
+                if ($newResource->equals($item)) {
+                    $itemQuantityModifier->modify($item, $item->getQuantity() + $newResource->getQuantity());
+                    $isItemInCart = true;
+
+                    break;
+                }
+            }
+
+            if (!$isItemInCart) {
+                $cart->addItem($newResource);
```

Comment:
But, we can extract it to private function, where you would be able to use an early return if condition is fulfilled, or just create a new resource.